core/src/aligned_polling_service.rs

```diff
@@ -9,7 +9,78 @@ use ethers::{
     signers::{Signer, Wallet},
     types::{Address, U256},
 };
-use log::info;
+use log::{error, info};
+
+use crate::proof::state_proof::{MinaStateProof, MinaStatePubInputs};
+
+/// Submits a Mina state proof to Aligned's batcher and waits until the batch is verified.
+#[allow(clippy::too_many_arguments)]
+pub async fn submit_state_proof(
+    proof: &MinaStateProof,
+    pub_input: &MinaStatePubInputs,
+    chain: &Chain,
+    proof_generator_addr: &str,
+    batcher_addr: &str,
+    batcher_eth_addr: &str,
+    eth_rpc_url: &str,
+    wallet: Wallet<SigningKey>,
+    save_proof: bool,
+) -> Result<AlignedVerificationData, String> {
+    let proof = bincode::serialize(proof)
+        .map_err(|err| format!("Failed to serialize state proof: {err}"))?;
+    let pub_input = bincode::serialize(pub_input)
+        .map_err(|err| format!("Failed to serialize public inputs: {err}"))?;
+
+    if save_proof {
+        std::fs::write("./protocol_state.pub", &pub_input).unwrap_or_else(|err| {
+            error!("{}", err);
+            process::exit(1);
+        });
+        std::fs::write("./protocol_state.proof", &proof).unwrap_or_else(|err| {
+            error!("{}", err);
+            process::exit(1);
+        });
+    }
+
+    let proof_generator_addr =
+        Address::from_str(proof_generator_addr).map_err(|err| err.to_string())?;
+
+    let verification_data = VerificationData {
+        proving_system: ProvingSystemId::Mina,
+        proof,
+        pub_input: Some(pub_input),
+        verification_key: None,
+        vm_program_code: None,
+        proof_generator_addr,
+    };
+
+    let wallet_address = wallet.address();
+    let nonce = get_nonce(eth_rpc_url, wallet_address, batcher_eth_addr).await?;
+
+    let proof_type = match verification_data.proving_system {
+        ProvingSystemId::Mina => "Mina Proof of State",
+        ProvingSystemId::MinaAccount => "Mina Proof of Account",
+        _ => return Err("Tried to submit a non Mina proof".to_string()),
+    };
```

Comment:
`proof_type` would always be `ProvingSystemId::Mina` because of line 49.

Reply:
good catch!